### ISSUE TITLE
Fedora:: Running "podman pull" during a "docker build" (i.e. within it's Dockerfile(5)) fails with:: "cannot clone: Operation not permitted; Error: cannot re-exec process" ...

### ISSUE THREAD

**Issue by nmvega** [containers/podman#9843]
Dear Friends:

`GUEST: Fedora-33` -- docker-ce image that I'm trying to build.
`HOST:  Fedora-30` -- There's a reason I need Fedora-30 here (I don't like the mismatch either). LoL

I successfully built a BASE `docker-ce` image where, among other items, I installed the `podman v3.0.1 ` ecosystem via `Fedora` RPMs. `FROM` that BASE `docker-ce` image, I'm trying to build a `CHILD` image during which I run `podman pull [...]` commands at build-time, as a `non-ROOT` user.

Thus when I run:

`root@HOST# docker build --tag acme/acme-foo:1.0 -f ./Dockerfile .`

when the `build` reaches the following statements:
```
FROM ${BASE_CONTAINER}  # Which has podman(1) installed.
[ ... snip ... ]
USER ${JANEDOE_UID}
RUN podman pull docker.io/library/mongo:latest
[ ... snip ... ]
```
it fails with the following message:

```
[ ... snip ... ]
cannot clone: Operation not permitted
Error: cannot re-exec process
root@HOST#
```

In fact, this failure doesn't only happen when building the `CHILD` image. When I `interactively` run the `BASE` image, I get the same error, as shown now:

```
root@HOST# docker run -it --user janedoe --workdir /home/janedoe <base-image-id> /bin/bash

janedoe@GUEST$ id
uid=1000(janedoe) gid=100(users) groups=100(users)

janedoe@GUEST$ pwd
/home/janedoe

janedoe@GUEST$ podman info
cannot clone: Operation not permitted
Error: cannot re-exec process

janedoe@GUEST$
janedoe@GUEST$ exit
root@HOST# 
```

Any ideas friends? Thank you in advance! =:)

**Comment by mheon** [comment]
@rhatdan Looks like a podman-in-container

**Comment by nmvega** [comment]
@mheon @rhatdan
That's correct. :relaxed: Thank you.

**Comment by rhatdan** [comment]
In order to run podman inside of Docker, you need to run a privileged container.  There are potentially other ways to do this a little more locked down, but for now, just run --privileged.

**Comment by rhatdan** [comment]
If all you want to do is build images using Containerfile/Dockerfile, I would recommend you use buildah.

**Comment by nmvega** [comment]
> If all you want to do is build images using Containerfile/Dockerfile, I would recommend you use buildah.
> https://developers.redhat.com/blog/2019/08/14/best-practices-for-running-buildah-in-a-container/

Good morning!

Agreed: I simply forgot to mention that when I run a fully `-d detached` container (i.e. not with `-it`), I do indeed also launch it with the `--privileged` option. `podman(1)` fully works in that case. And for the interactive case previously mentioned (i.e. with `-it`), I simply forgot to supply that same `--privileged` option. The scenario works, too. That was a red herring (my fault).

So we're back to my original `docker build [ ... ]` problem,  where the `--privileged` option isn't available. `podman(1) pull` statements inside the `Dockerfile` being built, causes the error in the `subject line`.

To be honest, I've never used `buildah`. =:) I will read more about it. But to review, here's what I'm doing:

1. On the bare-metal host which is running (and requires) the `docker-ce` engine, I need to build the docker image. I'm using `docker build [ ... ]` to do that, but running into the issue described because `podman pull [...]` statements are inside the `Dockerfile`.

2. Next, that built image will (later) be launched by an application using the docker `API`, which I have no control over. Meaning, it will require a `docker` compatible image as well as use the `docker-ce` engine, whether I build said image with `docker(1) build` or, say, `buildah(1) [... options ...]` or whatever else).

Are you suggesting that on my bare-metal host I substitute `docker build [ ... ]` with `buildah [ ... options ... ]`? Will that overcome my issue?

Thank you very much again!

**Comment by nmvega** [comment]
> In order to run podman inside of Docker, you need to run a privileged container. There are potentially other ways to do this a little more locked down, but for now, just run --privileged.

@rhatdan Good morning!

Agreed: I simply forgot to mention that when I run a fully `-d detached` container (i.e. not with `-it`), I do indeed also launch it with the `--privileged` option. `podman(1)` fully works in that case. And for the interactive case previously mentioned (i.e. with `-it`), I simply forgot to supply that same `--privileged` option. The scenario works, too. My fault (red herring).

So we're back to my original `docker build [ ... ]` problem,  where the `--privileged` option isn't available. `podman(1) pull` statements inside the `Dockerfile` being built, causes the error in the `subject line`.

To review, here's what I'm doing:

1. On the bare-metal host which is running (and requires) the `docker-ce` engine, I need to build the docker image. I'm using `docker build [ ... ]` to do that, but running into the issue described because `podman pull [...]` statements are inside the `Dockerfile`.

2. Next, that built image will (later) be launched by an application using the docker `API`. Meaning, it will expect a `docker` compatible image, in the classic repository location, as well as using the `docker-ce` engine

So I need to use the classic `docker build [ ... ]` command. `buildah` wouldn't suffice here, sadly, since it write to a different repository.

I hope that helps clarify. Thank you again!

**Comment by rhatdan** [comment]
Podman and Buildah will need mount/clone syscalls to run. Docker prevents these by default. So you at least needs Docker's seccomp filters modified in order to run the containers.  

You will also need at least CAP_SETUID and CAP_SETGID within the container in order to setup a user namespace.

**Comment by nmvega** [comment]
So for the `docker build` part, something like this:

`docker build --security-opt seccomp=unconfined [ ... ]`

That's what I found with a quick web search. I'll have to separately search for how to do second part: `CAP_SETUID` and `CAP_SETGID`.

Alternatively, perhaps it `could` be better to build the images (the `parent` and `child` images) using `buildah`, and then find a way to `copy` the resulting images to the `docker-ce` registry. Might that work? Is there a `copy/migration` command?

Sorry. I didn't intend this to be a tutorial.

**Comment by rhatdan** [comment]
You can use buildah push to push directly into a Docker daemon.

buildah push IMAGE docker-daemon:IMAGE

**Comment by nmvega** [comment]
@rhatdan Good morning. Thank you for that. I'll try this route. :)

**Comment by nmvega** [comment]
Hi @rhatdan Sorry for the delay. I'm still working on trying the `buildah(1)` route. I first ran into the `Dockerfile COPY` command issue that an earlier version of `buildah(1)` had, so I needed to stop and upgrade my `Fedora-30` servers (plural) to `Fedora-33` in order to upgrade the version of `buildah(1)`. That resolved the `Dockerfile COPY` issue.

Sadly, I just running into [this issue](https://github.com/containers/buildah/issues/1888), where I posted my variation of it. So I'm still plugging away at this.
=:)
